src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs

```diff
@@ -0,0 +1,517 @@
+                }
+                
+                if (parameter.Default != null)
+                {
+                    updatedDefaults.Add(parameter.Name, parameter.Default);
+                }
+
+                if (!updatedConstraints.TryGetValue(parameter.Name, out var parameterConstraints) &&
+                    parameter.Constraints.Count > 0)
+                {
+                    parameterConstraints = new List<RoutePatternConstraintReference>();
+                    updatedConstraints.Add(parameter.Name, parameterConstraints);
+                }
+
+                if (parameter.Constraints.Count > 0)
+                {
+                    parameterConstraints.AddRange(parameter.Constraints);
+                }
+
+                return ParameterPartCore(
+                    parameter.Name,
+                    @default,
+                    parameter.ParameterKind,
+                    (IEnumerable<RoutePatternConstraintReference>)parameterConstraints ?? Array.Empty<RoutePatternConstraintReference>());
+            }
+        }
+
+        public static RoutePatternPathSegment Segment(IEnumerable<RoutePatternPart> parts)
+        {
+            if (parts == null)
+            {
+                throw new ArgumentNullException(nameof(parts));
+            }
+
+            return SegmentCore(parts);
+        }
+
+        public static RoutePatternPathSegment Segment(params RoutePatternPart[] parts)
+        {
+            if (parts == null)
+            {
+                throw new ArgumentNullException(nameof(parts));
+            }
+
+            return SegmentCore(parts);
+        }
+
+        private static RoutePatternPathSegment SegmentCore(IEnumerable<RoutePatternPart> parts)
+        {
+            return new RoutePatternPathSegment(parts.ToArray());
+        }
+
+        public static RoutePatternLiteralPart LiteralPart(string content)
+        {
+            if (string.IsNullOrEmpty(content))
+            {
+                throw new ArgumentException(Resources.Argument_NullOrEmpty, nameof(content));
+            }
+
+            if (content.IndexOf('?') >= 0)
+            {
+                throw new ArgumentException(Resources.FormatTemplateRoute_InvalidLiteral(content));
+            }
+
+            return LiteralPartCore(content);
+        }
+
+        private static RoutePatternLiteralPart LiteralPartCore(string content)
+        {
+            return new RoutePatternLiteralPart(content);
+        }
+
+        public static RoutePatternSeparatorPart SeparatorPart(string content)
+        {
+            if (string.IsNullOrEmpty(content))
+            {
+                throw new ArgumentException(Resources.Argument_NullOrEmpty, nameof(content));
+            }
+
+            return SeparatorPartCore(content);
+        }
+
+        private static RoutePatternSeparatorPart SeparatorPartCore( string content)
+        {
+            return new RoutePatternSeparatorPart( content);
```

Comment:
nit: extra space

Reply:
wierd

Reply:
I wonder if the editor did that when I removed the `rawText` parameter